src/contract.rs

```diff
@@ -1,151 +1,314 @@
 use cosmwasm_std::{
-    to_binary, Api, Binary, Env, Extern, HandleResponse, InitResponse, MessageInfo, Querier,
+    attr, Api, Binary, Env, Extern, HandleResponse, HumanAddr, InitResponse, MessageInfo, Querier,
     StdResult, Storage,
 };
 
 use crate::error::ContractError;
-use crate::msg::{CountResponse, HandleMsg, InitMsg, QueryMsg};
-use crate::state::{config, config_read, State};
+use crate::helper::extract_funding_coin;
+use crate::msg::{HandleMsg, InitMsg, QueryMsg};
+use crate::state::{create_proposal, Config, Proposal, Vote, CONFIG, PROPOSALS, VOTES};
 
 // Note, you can use StdResult in some functions where you do not
 // make use of the custom errors
 pub fn init<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
-    _env: Env,
+    env: Env,
     info: MessageInfo,
     msg: InitMsg,
-) -> StdResult<InitResponse> {
-    let state = State {
-        count: msg.count,
-        owner: deps.api.canonical_address(&info.sender)?,
+) -> Result<InitResponse, ContractError> {
+    msg.validate(env, info)?;
+
+    let cfg = Config {
+        admin: msg.admin,
+        create_proposal_whitelist: msg.create_proposal_whitelist,
+        vote_proposal_whitelist: msg.vote_proposal_whitelist,
+        voting_period: msg.voting_period,
+        proposal_period: msg.proposal_period,
+        coin_denom: msg.coin_denom,
     };
-    config(&mut deps.storage).save(&state)?;
+    CONFIG.save(&mut deps.storage, &cfg)?;
 
     Ok(InitResponse::default())
 }
 
 // And declare a custom Error variant for the ones where you will want to make use of it
 pub fn handle<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
-    _env: Env,
+    env: Env,
     info: MessageInfo,
     msg: HandleMsg,
 ) -> Result<HandleResponse, ContractError> {
     match msg {
-        HandleMsg::Increment {} => try_increment(deps),
-        HandleMsg::Reset { count } => try_reset(deps, info, count),
+        HandleMsg::CreateProposal {
+            title,
+            description,
+            metadata,
+            fund_address,
+        } => try_create_proposal(deps, env, info, title, description, metadata, fund_address),
+        HandleMsg::VoteProposal { proposal_id } => try_vote_proposal(deps, env, info, proposal_id),
+        HandleMsg::TriggerDistribution { .. } => Ok(HandleResponse::default()),
```

Comment:
```suggestion
        HandleMsg::TriggerDistribution { .. } => panic!("unimplemented")
```
or better
```suggestion
        HandleMsg::TriggerDistribution { proposal_id } => try_trigger_distribution(deps, env, info, proposal_id),
```
and put the `panic!` inside `try_trigger_distribution()` (along with a `//TODO`, maybe).